test/serve-asset-middleware-test.js

```diff
@@ -0,0 +1,206 @@
+var RSVP = require('rsvp');
+var serveAssetMiddleware = require('./../lib/index').serveAssetMiddleware;
+var watcherMiddleware = require('./../lib/index').watcherMiddleware;
+var fixture = require('./helpers/fixture-path');
+var TestHTTPServer = require('./helpers/test-http-server');
+
+describe('serve-middleware', function() {
+  describe('watcher resolves correctly', function() {
+    var server;
+
+    afterEach(function() {
+      server.stop();
+      server = null;
+    })
+
+    it('serves the given file if found', function() {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('basic-file')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+      var serveMiddleware = serveAssetMiddleware();
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(serveMiddleware);
+
+      return server.start()
+        .then(function(info) {
+          return server.request('/index.html', {
+            info: info
+          });
+        })
+        .then(function(content) {
+          expect(content).to.match(/This is broccoli middleware page/);
+        });
+    });
+
+    it('no response headers are set & it calls the next middleware when file is not found', function(done) {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('basic-file')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+
+      var serveAssetMiddlewareFn = serveAssetMiddleware();
+      var wrapperMiddleware = function(req, resp/*, next*/) {
+        serveAssetMiddlewareFn(req, resp, function() {
+          // assert response headers
+          expect(resp.getHeader('Cache-Control')).to.be.equal('private, max-age=0, must-revalidate');
+          expect(resp.getHeader('Last-Modified')).to.not.be.ok;
+          expect(resp.getHeader('Content-Type')).to.not.be.ok;
+          done();
+        });
+      }
+
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(wrapperMiddleware);
+
+      server.start()
+        .then(function(info) {
+          return server.request('/non-existent-file', {
+            info: info
+          });
+        });
+    });
+
+    it('bypasses middleware if request is a directory and autoIndex is set to false', function(done) {
+      var watcher = RSVP.Promise.resolve({
+        'directory': fixture('no-index')
+      });
+
+      var middleware = watcherMiddleware(watcher, {
+        autoIndex: false
+      });
+
+      var serveAssetMiddlewareFn = serveAssetMiddleware();
+      var wrapperMiddleware = function(req, resp /*next*/) {
+        serveAssetMiddlewareFn(req, resp, function() {
+          var broccoliHeader = req.headers['x-broccoli'];
+          expect(broccoliHeader.filename).to.equal('');
+          done();
+        });
+      };
+
+      server = new TestHTTPServer(middleware);
+      server.addMiddleware(wrapperMiddleware);
+
```

Comment:
The wrapperMiddleware creates a closure to allow doing assertions. I could otherwise just create a another middleware to do the assertions after serveAssetMiddleware is finished executing.